raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.ShortNode;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.raft.generated.QuorumStateData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.util.OptionalInt;
+
+/**
+ * Local file based quorum state store. It takes the JSON format of {@link QuorumStateData}
+ * with an extra data version number as part of the data for easy deserialization.
+ *
+ * Example format:
+ * <pre>
+ * {"clusterId":"",
+ *   "leaderId":1,
+ *   "leaderEpoch":2,
+ *   "votedId":-1,
+ *   "appliedOffset":0,
+ *   "currentVoters":[],
+ *   "targetVoters":[],
+ *   "data_version":0}
+ * </pre>
+ * */
+public class FileBasedStateStore implements QuorumStateStore {
+    private static final Logger log = LoggerFactory.getLogger(FileBasedStateStore.class);
+
+    private final File stateFile;
+
+    private QuorumStateData cachedState;
+
+    static final String DATA_VERSION = "data_version";
+
+    public FileBasedStateStore(final File stateFile, OffsetAndEpoch logEndOffsetAndEpoch) throws IOException {
+        this.stateFile = stateFile;
+        cachedState = loadState(stateFile, logEndOffsetAndEpoch);
```

Comment:
I really think we're making this harder than it needs to be. The only time we call `readElectionState` is in `QuorumState.initialize`, so what is the benefit of this preloading and caching? It would be simpler to just let `readElectionState` read from the file and get rid of the cached state.

Also, why do we need to pass through `logEndOffsetAndEpoch`? This creates an awkward dependency chain with `ReplicatedLog`, which is outside of the normal initialization path. It seems the only reason is so that we can use it to initialize state after catching exceptions. However, the `readElectionState` method was intended to raise exceptions so that they could be handled generally. By moving the exception handling into `QuorumState`, then we no longer need to pass through `logEndOffsetAndEpoch`.

Reply:
That sounds reasonable, will have a try.